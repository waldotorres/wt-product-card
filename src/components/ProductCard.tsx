import styles from "../styles/styles.module.css";
import { useProduct  } from '../hooks/useProduct';
import React,{ createContext } from 'react';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCartHandlers } from '../interfaces/interfaces';


export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface ProductCardProps {
  product:Product,
  //children: ReactElement | ReactElement[],
  children:(args:ProductCartHandlers)=>JSX.Element,
  className?:string,
  style?:React.CSSProperties,
  onChange?: (args: onChangeArgs)=>void;
  value?:number,
  initialValues?:InitialValues
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: ProductCardProps  ) => {


  //
  const { counter, increaseBy, maxCount, isMaxCountReached, reset }=  useProduct({ onChange, product, value, initialValues });

  return (

    <Provider  value= { {counter, increaseBy, product, maxCount}  }  >
      <div className={ `${styles.productCard}  ${className} `} style={style} >

        { children( {
          maxCount,
          count: counter,
          isMaxCountReached,
          product,
          increaseBy,
          reset
        }) }
      </div>

    </Provider>
 
  )
}


// ProductCard.Title =
// ProductCard.Buttons =
// ProductCard.Image =

 //   <div className= { styles.productCard } >

  //   {/* <img className={ styles.productImg } src="./coffee-mug.png" alt="Coffe Mug" /> */}
  //   {/* <img className={ styles.productImg } src= { noImage }  alt="Coffe Mug" /> */}

  //   {/* <span className={ styles.productDescription }>Coffe Mug</span> */}
  //   {/* <div className={ styles.buttonsContainer }>
  //       <button className= { styles.buttonMinus }
  //               onClick ={ ()=>increaseBy(-1) } >-</button>
  //       <div className={ styles.countLabel }> { counter } </div>
  //       <button className={ styles.buttonAdd }
  //               onClick ={ ()=>increaseBy(1) }>+</button>
  //   </div> */}
  // </div>