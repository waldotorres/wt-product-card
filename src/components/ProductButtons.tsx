
import styles from "../styles/styles.module.css";
import { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import * as React from 'react';


export interface ProductButtonsProps {
        className?:string,
        style?:React.CSSProperties
}

export const ProductButtons = ( {className, style}:ProductButtonsProps ) => {

    const { increaseBy, counter, maxCount=-1 } = useContext( ProductContext );

    
//console.log({counter ,  maxCount})

    const shouldDisabled = useCallback(
        ()=>( (counter >= maxCount)?true:false  ) ,
      [counter, maxCount],
    )
    

        //console.log(shouldDisabled())

    return (
    
    <div   className={ `${styles.buttonsContainer} ${className}` } 
           style={style}>
                
        <button className= { styles.buttonMinus }
                onClick ={ ()=>increaseBy(-1) } >-</button>
        <div className={ styles.countLabel }> { counter } </div>
        <button className={ `${styles.buttonAdd} ${ shouldDisabled() && styles.disabled } ` }
                onClick ={ ()=>increaseBy(1) }>+</button>
    </div>

  )
}
