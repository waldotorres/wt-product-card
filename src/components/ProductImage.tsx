import React, { useContext } from "react";
import styles from "../styles/styles.module.css";
import { ProductContext } from './ProductCard';

import noImage from '../assets/no-image.jpg';

export interface ProductImageProps {
  img?:string,
  className?:string,
  style?:React.CSSProperties

}

export const ProductImage = ({ img, className, style }:ProductImageProps) => {

  const { product }= useContext(ProductContext);
  let imgToShow = '';

  if( img ){
    imgToShow = img;
  }
  else if(product.img){
    imgToShow = product.img
  }
  else{
    imgToShow = noImage
  }

  return (
    
    <img  className={ `${styles.productImg} ${ className}` } src= { imgToShow }  
          style={style}
          alt="Coffe Mug" />
  )
}
