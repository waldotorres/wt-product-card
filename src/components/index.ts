import { ProductCard as ProductCardHOC } from "./ProductCard";
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductTitle } from './ProductTitle';
import { ProductContextHOCProps } from "../interfaces/interfaces";

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from './ProductImage';
export { ProductTitle } from "./ProductTitle";


export const ProductCard:ProductContextHOCProps = Object.assign( ProductCardHOC, 
                                {   Image: ProductImage, 
                                    Buttons: ProductButtons, 
                                    Title: ProductTitle } )

export default ProductCard;                                    