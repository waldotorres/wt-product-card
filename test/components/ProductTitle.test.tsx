import React from "react";
import renderer from "react-test-renderer";
import { ProductTitle } from '../../src/components/ProductTitle';


describe('ProductTitle', () => {
   
    test( 'debe mostraar el componente con el titulo personalizado', ()=>{
        const wrapper = renderer.create (
            <ProductTitle title ="Custom product"   />

            
        )

        //console.log(wrapper.toJSON());
        expect( wrapper.toJSON()).toMatchSnapshot();
    })


});
