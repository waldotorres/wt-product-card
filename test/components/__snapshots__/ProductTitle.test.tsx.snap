// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostraar el componente con el titulo personalizado 1`] = `
<span
  className="productDescription undefined "
>
   
  Custom product
   
</span>
`;
